src/poetry/utils/env/python/manager.py

```diff
@@ -0,0 +1,197 @@
+    def executable(self) -> Path:
+        return cast(Path, self._python.interpreter)
+
+    @property
+    def version(self) -> Version:
+        return Version.parse(str(self._python.version))
+
+    @cached_property
+    def patch_version(self) -> Version:
+        return Version.from_parts(
+            major=self.version.major,
+            minor=self.version.minor,
+            patch=self.version.patch,
+        )
+
+    @cached_property
+    def minor_version(self) -> Version:
+        return Version.from_parts(major=self.version.major, minor=self.version.minor)
+
+    @classmethod
+    def get_active_python(cls) -> Python | None:
+        """
+        Fetches the active Python interpreter from available system paths or falls
+        back to finding the first valid Python executable named "python".
+
+        An "active Python interpreter" in this context is an executable (or a symlink)
+        with the name `python`. This is done so to detect cases where pyenv or
+        alternatives are used.
+
+        This method first uses the `ShutilWhichPythonProvider` to detect Python
+        executables in the path. If no interpreter is found using, it attempts
+        to locate a Python binary named "python" via the `findpython` library.
+
+        :return: An instance representing the detected active Python,
+                 or None if no valid environment is found.
+        """
+        for python in ShutilWhichPythonProvider().find_pythons():
+            return cls(python=python)
+
+        # fallback to findpython, restrict to finding only executables
+        # named "python" as the intention here is just that, nothing more
+        if python := findpython.find("python"):
+            return cls(python=python)
+
+        return None
+
+    @classmethod
+    def from_executable(cls, path: Path | str) -> Python:
+        try:
+            return cls(python=findpython.PythonVersion(executable=Path(path)))
+        except (FileNotFoundError, NotADirectoryError, ValueError):
+            raise ValueError(f"{path} is not a valid Python executable")
+
+    @classmethod
+    def get_system_python(cls) -> Python:
+        """
+        Creates and returns an instance of the class representing the Poetry's Python executable.
+        """
+        return cls(
+            python=findpython.PythonVersion(
+                executable=Path(sys.executable),
+                _version=packaging.version.Version(
+                    ".".join(str(v) for v in sys.version_info[:3])
+                ),
+            )
+        )
+
+    @classmethod
+    def get_by_name(cls, python_name: str) -> Python | None:
+        if Path(python_name).exists():
+            with contextlib.suppress(AssertionError, ValueError):
+                # if it is a path try assuming it is an executable
+                python = cls.from_executable(python_name)
+                assert python.version is not None
+                return cls.from_executable(python_name)
```

Comment:
Why we need the `assert`? And even if it is required, can't we return `python` instead of running `from_executabl` again? :thinking:

Reply:
We need the assert because sometimes given path might not be a valid python and we cannot be sure unless we execute something. Version allows this and ensures it's a Python we can get version from.

And yes we can rerun the instance. It just got missed in the refactor.

Reply:
Yeah, that assert was an artifact of the first iteration. Removed it now and changed it to suppressing just the ValueError. This should be sufficient for our needs.